### PR TITLE
Taxon (category) deletion

### PR DESCRIPTION
**Pull request by SagarKarwande**
[delivers #161198500]

## Why?
User should be able to delete a particular category from the taxonomy.

## This change addresses the need by:
`core`
--------
- Added following API for Product 
  - Delete products by category
  - Get products by category

`admin_app`
----------------
- Taxon delete functionality on Taxonomy page. 

[delivers #161198500]

## Checklist

- [ ] I have read [CONTRIBUTING.md][contributing].
- [ ] My code follows the [style guidelines][contributing] of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] My code does not generate any (new) [`credo`][credo] and compile-time warnings.
- [x] I have updated the documentation wherever necessary.
- [x] I have added tests to cover my changes.

[contributing]: https://github.com/aviabird/snitch/CONTRIBUTING.md
[credo]: https://github.com/rrrene/credo

### CODE REVIEW

File: apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex [hunk 1]
```diff
@@ -275,7 +287,47 @@ defmodule Snitch.Domain.Taxonomy do
...
+        {:error, :not_found}
+    end
+  end
+
+  defp do_delete_products(taxon_id) do
+    categories =
+      get_all_children_and_self(taxon_id)
+      |> Enum.map(& &1.id)
+
+    product_delete_query =
```
Comment (ashish173) [inline apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex]: This query should come from product model. We want to fetch all the products of a category. This is already implemented in the codebase.

File: apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex [hunk 2]
```diff
@@ -275,7 +287,47 @@ defmodule Snitch.Domain.Taxonomy do
...
+        |> Multi.run(:category, fn _ -> do_delete_taxon(taxon) end)
+        |> Repo.transaction()
+
+      nil ->
+        {:error, :not_found}
+    end
+  end
+
+  defp do_delete_products(taxon_id) do
+    categories =
```
Comment (ashish173) [inline apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex]: can be extracted to product model. with success and error return. which will let the multi move ahead.

File: apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex [hunk 3]
```diff
@@ -275,7 +287,47 @@ defmodule Snitch.Domain.Taxonomy do
...
+    product_delete_query =
+      from(p in Product,
+        where: p.taxon_id in ^categories
+      )
+
+    total_products = from(p in product_delete_query, select: count(p.id)) |> Repo.one()
+
+    {update_products_count, t} =
+      Repo.update_all(product_delete_query,
+        set: [state: "deleted", deleted_at: NaiveDateTime.utc_now(), taxon_id: nil]
```
Comment (ashish173) [inline apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex]: @SagarKarwande @aviacommerce/snitch-core this is a very important thing. We CANNOT have the logic of product delete in more than 1 place. The logic to set the `delete_at` to current_time. 

This is violating the `DRY` principle. Please make sure not do this type of things.

File: apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex [hunk 4]
```diff
@@ -169,6 +169,18 @@ defmodule Snitch.Domain.Taxonomy do
     Repo.all(from(taxon in Taxon, where: taxon.parent_id == ^taxon_id))
```
Comment (ashish173) [inline apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex]: @SagarKarwande refactor this to use `as_nested_set` function.

Comment (SagarKarwande) [reply apps/snitch_core/lib/core/domain/taxonomy/taxonomy.ex]: Done

**Pull request merged**